dynamic_programming/max_product_subarray.py

```diff
@@ -0,0 +1,50 @@
+def max_product_subarray(numbers: list[int]) -> int:
+    """
+    Returns the maximum product that can be obtained by multiplying a
+    contiguous subarray of the given integer list `nums`.
+
+    Example:
+    >>> max_product_subarray([2, 3, -2, 4])
+    6
+    >>> max_product_subarray([-2, 0, -1])
+    0
+    >>> max_product_subarray([2, 3, -2, 4, -1])
+    48
+    >>> max_product_subarray([2, 3, -2, 4.5, -1])
+    54
+    >>> max_product_subarray([-1])
+    -1
+    >>> max_product_subarray([])
+    0
+    >>> max_product_subarray("ABC")
+    0
+    >>> max_product_subarray("")
+    0
+    >>> max_product_subarray(None)
+    0
+    """
+    if numbers is None or not isinstance(numbers, list):
+        return 0
+
+    n = len(numbers)
+
+    if n == 0:
+        return 0
+
+    if not all(isinstance(x, (int, float)) for x in numbers):
+        return 0
```

Comment:
It is the Pythonic thing to do to raise a TypeError if the caller passes in garbage data.